Content.Server/Animals/Systems/UdderSystem.cs

```
@@ -0,0 +1,153 @@
+                return;
+            }
+
+            udder.BeingMilked = true;
+
+            var doargs = new DoAfterEventArgs(userUid, 5, default, uid)
+            {
+                BreakOnUserMove = true,
+                BreakOnDamage = true,
+                BreakOnStun = true,
+                BreakOnTargetMove = true,
+                MovementThreshold = 1.0f,
+                TargetFinishedEvent = new MilkingFinishedEvent(userUid, containerUid),
+                TargetCancelledEvent = new MilkingFailEvent()
+            };
+
+            _doAfterSystem.DoAfter(doargs);
+        }
+
+        private void OnMilkingFinished(EntityUid uid, UdderComponent udder, MilkingFinishedEvent ev)
+        {
+            udder.BeingMilked = false;
+
+            if (!_solutionContainerSystem.TryGetSolution(uid, udder.TargetSolutionName, out var solution))
+                return;
+
+            if (!_solutionContainerSystem.TryGetRefillableSolution(ev.ContainerUid, out var targetSolution))
+                return;
+
+            var user = EntityManager.GetEntity(ev.UserUid);
+
+            var quantity = solution.TotalVolume;
+            if(quantity == 0)
+            {
+                udder.Owner.PopupMessage(user, Loc.GetString("udder-system-dry"));
+                return;
+            }
+
+            if (quantity > targetSolution.AvailableVolume)
+                quantity = targetSolution.AvailableVolume;
+
+            var split = _solutionContainerSystem.SplitSolution(uid, solution, quantity);
+            _solutionContainerSystem.TryAddSolution(ev.ContainerUid, targetSolution, split);
+
+            var container = EntityManager.GetEntity(ev.ContainerUid);
+            udder.Owner.PopupMessage(user, Loc.GetString("udder-system-success", ("amount", quantity), ("target", container)));
+        }
+
+        private void OnMilkingFailed(EntityUid uid, UdderComponent component, MilkingFailEvent ev)
+        {
+            //TODO: fail PopupMessage?
```

```suggestion
```
probably not really needed, doafter already fails visibly